web/packages/teleterm/src/mainProcess/mainProcess.ts

```diff
@@ -191,6 +193,15 @@ export default class MainProcess {
     const { binaryPath, homeDir } = this.settings.tshd;
     this.logger.info(`Starting tsh daemon from ${binaryPath}`);
 
+    // spawn might either fail immediately by throwing an error or cause the error event to be emitted
```

Comment:
Theoretically, we could unify this behavior by checking if the process [has `.pid` set](https://nodejs.org/api/child_process.html#subprocesspid).
I tired it and seems to work correctly.

```ts
    this.tshdProcess = spawn(
      binaryPath,
      ['daemon', 'start', ...this.getTshdFlags()],
      {
        stdio: 'pipe', // stdio must be set to `pipe` as the gRPC server address is read from stdout
        windowsHide: true,
        env: {
          ...process.env,
          TELEPORT_HOME: homeDir,
          [TSH_AUTOUPDATE_ENV_VAR]: TSH_AUTOUPDATE_OFF,
        },
      }
    );
    if (!this.tshdProcess.pid) {
      return new Promise((_, reject) => {
        this.tshdProcess.on('error', reject);
      });
    }
```

Reply:
Hmm, I don't understand why a promise is returned there since `initTshd` is not async. The docs for the pid property says "If the child process fails to spawn due to errors, then the value is undefined and error is emitted." so it looks like this should be built-in?

We could probably add a try/catch around spawn and on error just swallow it and let the error event fire I suppose. But also I'm not sure if it's strictly necessary, the stack trace for sync errors is much better than the one from the error event. Not to mention that funny behavior of Node.js event emitter, where the error event is throw as an error instead if there are no listeners for the error event at the time of emitting it.

Reply:
>Hmm, I don't understand why a promise is returned there since initTshd is not async. 

Ah, I forgot to mention that `initTshd` should become async  :)
It's a bit weird that the error is only available in the `error` event listener instead of being thrown from `spawn`, while `pid: undefined` is set immediately. 

>We could probably add a try/catch around spawn and on error just swallow it and let the error event fire I suppose. But also I'm not sure if it's strictly necessary, the stack trace for sync errors is much better than the one from the error event.

A better stack trace sounds like a good reason to keep it as is.